sources/web/datalab/idleTimeout.ts

```diff
@@ -0,0 +1,256 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/// <reference path="common.d.ts" />
+
+import childProcess = require('child_process');
+import http = require('http');
+import logging = require('./logging');
+import net = require('net');
+import querystring = require('querystring');
+import url = require('url');
+import userManager = require('./userManager');
+
+const idleCheckIntervalSeconds = 5;    // Seconds to wait between idle checks
+
+let idleTimeoutEnabled = true;  // Allow for explicit disable without clearing other values
+let idleTimeoutSeconds = 0;   // Shutdown after being idle this long; disabled if 0 or NaN
+let lastUserActivity: number;  // The epoch, in seconds, of the last user activity
+let shutdownCommand = '';
+
+export function initAndStart(appSettings: common.Settings) {
+  init(appSettings);
+  reset();
+  startChecker();
+}
+
+export function init(appSettings: common.Settings) {
+  shutdownCommand = appSettings.idleTimeoutShutdownCommand;
+  if (!shutdownCommand) {
+    shutdownCommand = process.env.DATALAB_SHUTDOWN_COMMAND;
+  }
+  if (shutdownCommand) {
+    idleTimeoutSeconds = appSettings.idleTimeoutSeconds;
+    logging.getLogger().debug('idleTimeoutSeconds from settings: ' + idleTimeoutSeconds);
+    if (!idleTimeoutSeconds) {
+      idleTimeoutSeconds = parseInterval(process.env.DATALAB_IDLE_TIMEOUT);
+    }
+    logging.getLogger().debug('Idle timeout set to ' + idleTimeoutSeconds + ' seconds');
+    if (! (idleTimeoutSeconds > 0)) {
+      logging.getLogger().debug('No idle timeout value, idle timeout is disabled');
+    }
+  } else {
+    idleTimeoutSeconds = NaN;   // No shutdown command available
+    logging.getLogger().debug('No shutdown command, idle timeout is disabled');
+  }
+}
+
+// Parse a string like '30s' or '1h 15m 5s' and return seconds.
+// If no input string, or unrecognized stuff, returns NaN.
+function parseInterval(str: string) {
+  if (!str) {
+    return NaN;
+  }
+  let total = 0;
+  const regAndMults = [
+    [ /(\d+)\s*(seconds?|s)/, 1 ],
+    [ /(\d+)\s*(minutes?|m)/, 60 ],
+    [ /(\d+)\s*(hours?|h)/, 60 * 60 ],
+    [ /(\d+)\s*(days?|d)/, 60 * 60 * 24 ],
+  ];
+  regAndMults.forEach(rm => {
+    const r : RegExp = rm[0] as RegExp;
+    const m : number = rm[1] as number;
+    const uStr = str.match(r);
+    str = str.replace(r, '');
+    if (uStr) { total += parseInt(uStr[1]) * m; }
+  });
+  str = str.replace(/\s*/, '');
+  if (str) {
+    logging.getLogger().debug('garbage left over in interval string: ' + str);
+    return NaN;
+  }
+  return total;
+}
+
+export function resetBasedOnPath(path: string) {
```

Comment:
looks like this is only called for unchecked requests, but it only does any work if the path is timeout, which means the path was checked. is this on purpose maybe to catch changes to `shouldResetOnPath`?